static/js/composebox_typeahead.js

```diff
@@ -330,6 +331,11 @@ exports.tokenize_compose_str = function (s) {
         }
     }
 
+    const timestamp_index = s.indexOf('!time');
+    if (timestamp_index >= 0) {
+        return s.slice(timestamp_index);
+    }
```

Comment:
Why is this a totally unrelated block?  I don't understand why the timestamp stuff needs to be a unique thing different from how `#` or `@` works; it's confusing and feels not maintainable.

Reply:
Because the others use switch as:

```
const first_char = token[0];
switch(first_char) {
    case '#':
        // do something.
}
```

But we need to match the whole string.

Reply:
Hmm, not sure I follow -- can you explain in more detail?

Reply:
The `switch` construct above works on a single character at a time, and most other typeaheads behave that way: `#` for stream, `>` for topic, `@` for user mention, `/` for slash commands. Here, we want to trigger for specifically `!time (+ some optional text)`, so saying that just a simple `!` could be a timestamp trigger is too broad.